lib/Fregot/Builtins/Basics.hs

```diff
@@ -279,6 +281,15 @@ builtin_min = Builtin
         [] -> Value NullV  -- TODO(jaspervdj): Should be undefined.
         _  -> minimum (vals :: [Value])
 
+builtin_numbers_range :: Monad m => Builtin m
+builtin_numbers_range = Builtin
+    (Ty.number 🡒 Ty.number 🡒 Ty.out (Ty.arrayOf Ty.number)) $ pure $
+    \(Cons a (Cons b Nil)) ->
+      return $! case compare (a :: Int) (b :: Int) of
+        EQ -> [a] :: [Int]
+        LT -> [a..b]
+        GT -> [a, (a-1)..b]
```

Comment:
One really tiny nitpick: we could use [enumFromStepN](https://hackage.haskell.org/package/vector-0.12.1.2/docs/Data-Vector.html#g:11) to avoid building a temporary list.

Reply:
🤔 That doesn't allow the step to be negative, or rather, `enumFromStepN 10 (-1) 1 == [10]`. Am I missing something? We could throw in a `V.reverse`...

Reply:
Oh wait I've misunderstood `enumFromStepN`. It's the number of steps 🤦

Reply:
The third argument of `enumFromStepN` is the number of items in the vector, not the last element.  So e.g. `enumFromStepN 10 (-1) 5 == [10,9,8,7,6]`.  I think it was designed that way so you don't need to do the (possibly expensive?) equality check on `a`, but I could be wrong. 🤷

Reply:
Hm. I'm not sure if this better, but it's using `enumFromStepN` 😅

Reply:
I guess we traded a little readability for a little performance.  I'm okay with it.

Interestingly, it would perhaps be better if `range` was a relation (like `walk`), not a function.  That way you could do things like:

```rego
numbers.range(1, 100000, out)
is_prime(out)
...
```

Without ever constructing the full array in memory.

Reply:
Yup that would be nice.